pyproject.toml

```diff
@@ -1,3 +1,3 @@
 [build-system]
-requires = ["setuptools", "wheel", "setuptools_dso>=2.1", "epicscorelibs>=7.0.6.99.1.0"]
+requires = ["setuptools", "wheel", "setuptools_dso>=2.1", "epicscorelibs==7.0.6.99.1.0"]
```

Comment:
This strong tie to a very particular version of `epicscorelibs` rings an alarm bell for me.  What's going on here?  Surely we don't want this, do we?

Separately, the version number is in an alarmingly non-standard format!

Reply:
> This strong tie to a very particular version of epicscorelibs rings an alarm bell for me. What's going on here?

There is an ordering issue here. The build process is this:
- deps from pyproject.toml are installed
- `python setup.py bdist_wheel` is called to make a binary wheel, which is compiled against the exact version of epicscorelibs that has been installed, and records this fact in the wheel
- the wheel is uploaded to PyPI
- anyone using this version of the binary wheel must have this exact version of epicscorelibs installed too (which pip will do automatically)
- anyone using the sdist is free to have whatever version of epicscorelibs they want

The `>=` that was there originally was correct, which would mean that pythonSoftIOC was compiled against the latest available epicscorelibs. Unfortunately in the tests we require p4p, which is also compiled against a particular version of epicscorelibs. If a new version of epicscorelibs is released, but p4p is not, then pip will fail to install dependencies. This is why we are using `==`. Without ABI compliance for EPICS, I can't see a way around this, unless a release of p4p is made with every epicscorelibs release. Fortunately we don't care for aioca as we only need API compliance, not ABI compliance.

@mdavidsaver can you think of a way we can resolve this?

> Separately, the version number is in an alarmingly non-standard format!

The first part is the EPICS release, but not sure about the other bits. Maybe @mdavidsaver could comment on this too?

Reply:
> If a new version of epicscorelibs is released, but p4p is not

Sorry.  I've been sloppy about this as I'm in the (so far long) process of introducing another dependency [pvxslibs](https://pypi.org/project/pvxslibs/).  I'll try to straighten this out.  Though even if I do, there will always be some time when pypy.org has newer versions of parts of the dependency chain.  Maybe only minutes, but never zero.

> we only need API compliance, not ABI compliance.

Speaking strictly, API is important at build time (with a compiler), and ABI afterwards.  So in an ideal world, pyproject.toml is expressing the build time dependency (`BuildRequires` in an RPM .spec file), while `install_requires` in setup.py expresses the runtime dependency (plain `Requires` in a .spec).  As with a `.spec` runtime dependencies are partly generated.  so...

https://github.com/dls-controls/pythonSoftIOC/blob/3866562259982b6b019d3d5cb2b87b3c8ec98b01/setup.py#L108-L109

So if epicscorelibs `7.0.6.99.1.0` is installed, this will expand to:

```
epicscorelibs >=7.0.6.99.1.0, <7.0.6.99.2
```


> Separately, the version number is in an alarmingly non-standard format!

There are three pieces of information being encoded: upstream EPICS Base version (everything up to and including the `.99`), epicscorelibs ABI, and epicscorelibs revision.  So the version range being emitted above allows for later revisions which (hopefully) don't include any ABI change.

Reply:
> @mdavidsaver can you think of a way we can resolve this?

If the problem is only with GHA, you might be able to pass `--no-binary p4p` to pip when installing the dev dependencies prior to running tests.

Reply:
Thanks @mdavidsaver it looks like p4p 3.5.5 fixes the issue - I've restored the original epicscorelibs dependency in this commit and it seems to be working: https://github.com/dls-controls/pythonSoftIOC/pull/60/commits/adaa2868794c9f3281d159ba0e0d2cb427f4cd55